packages/atomic/src/components/common/layout/custom-render-controller.spec.ts

```diff
@@ -0,0 +1,212 @@
+import {beforeEach, describe, expect, it, vi} from 'vitest';
+import {buildFakeProduct} from '@/vitest-utils/testing-helpers/fixtures/headless/commerce/product';
+import type {AnyItem} from '../item-list/unfolded-item';
+import {
+  CustomRenderController,
+  type CustomRenderHost,
+  type CustomRenderOptions,
+} from './custom-render-controller';
+
+@customElement('test-custom-render-element')
+class TestCustomRenderElement extends LitElement implements CustomRenderHost {
+  addController = vi.fn();
+
+  getRootNode(): Node {
+    return this.shadowRoot || document;
+  }
+}
+
+describe('CustomRenderController', () => {
+  let mockElement: TestCustomRenderElement;
+  let mockOptions: CustomRenderOptions;
+  let controller: CustomRenderController;
+  let mockRootElement: HTMLElement;
+  let mockLinkContainer: HTMLElement;
+  let mockItemData: AnyItem;
+  let mockRenderingFunction: ReturnType<typeof vi.fn>;
+  let mockOnRenderComplete: ReturnType<typeof vi.fn>;
+
+  beforeEach(() => {
+    mockElement = new TestCustomRenderElement();
+    mockRootElement = document.createElement('div');
+    mockLinkContainer = document.createElement('div');
+    mockItemData = buildFakeProduct();
+
+    mockRenderingFunction = vi
+      .fn()
+      .mockReturnValue('<div>Custom render output</div>');
+    mockOnRenderComplete = vi.fn();
+
+    mockOptions = {
+      renderingFunction: vi.fn().mockReturnValue(mockRenderingFunction),
+      itemData: vi.fn().mockReturnValue(mockItemData),
+      rootElementRef: vi.fn().mockReturnValue(mockRootElement),
+      linkContainerRef: vi.fn().mockReturnValue(mockLinkContainer),
+      onRenderComplete: mockOnRenderComplete,
+    };
+  });
+
+  describe('#constructor', () => {
+    beforeEach(() => {
+      controller = new CustomRenderController(mockElement, mockOptions);
+    });
+
+    it('should initialize with provided parameters', () => {
+      expect(mockElement.addController).toHaveBeenCalledWith(controller);
+    });
+
+    it('should set default values for optional options', () => {
+      const minimalOptions = {
+        renderingFunction: vi.fn().mockReturnValue(mockRenderingFunction),
+        itemData: vi.fn().mockReturnValue(mockItemData),
+        rootElementRef: vi.fn().mockReturnValue(mockRootElement),
+      };
+
+      controller = new CustomRenderController(mockElement, minimalOptions);
+
+      expect(mockElement.addController).toHaveBeenCalledWith(controller);
+    });
+  });
+
+  describe('#hostConnected', () => {
+    beforeEach(() => {
+      controller = new CustomRenderController(mockElement, mockOptions);
+    });
+
+    it('should not throw when called', () => {
+      expect(() => controller.hostConnected()).not.toThrow();
+    });
+
+    it('should call render function execution after connection', () => {
+      controller.hostUpdated();
+      expect(mockRenderingFunction).toHaveBeenCalledTimes(1);
+    });
+
+    it('should reset state when #hostConnected is called', () => {
```

Comment:
```suggestion
    it('should reset state', () => {
```

implicit from the describe wrapper.